src/tasks/processBulkUpload.ts

```diff
@@ -85,7 +310,58 @@ export const processBulkUpload = async (
   let bulkUploadFile: FileResult;
   try {
     await updateBulkUploadStatus(bulkUpload.id, BulkUploadStatus.IN_PROGRESS);
-    bulkUploadFile = await downloadS3ObjectToTemporaryStorage(bulkUpload.sourceKey);
+    bulkUploadFile = await downloadS3ObjectToTemporaryStorage(
+      bulkUpload.sourceKey,
+      helpers.logger,
+    );
+    await assertBulkUploadCsvIsValid(bulkUploadFile.path);
+    const opportunity = await createOpportunityForBulkUpload();
+    const applicationForm = await createApplicationFormForBulkUpload(opportunity.id);
+    const applicationFormFields = await createApplicationFormFieldsForBulkUpload(
+      bulkUploadFile.path,
+      applicationForm.id,
+    );
+
+    const applicantEmailFieldIndex = await getApplicantEmailFieldIndexForBulkUpload(
+      bulkUploadFile.path,
+    );
+    const csvReadStream = fs.createReadStream(bulkUploadFile.path);
+    const parser = parse({
+      from: 2,
+    });
+    csvReadStream.pipe(parser);
+    let recordNumber = 0;
+    await parser.forEach(async (record: string[]) => {
+      recordNumber += 1;
+      const applicantEmail = record[applicantEmailFieldIndex];
+      if (applicantEmail === undefined) {
+        throw new Error('The applicant\'s email address must be provided.');
+      }
+      const applicant = await createOrLoadApplicantForBulkUploadCsvRecord(applicantEmail);
+      const proposal = await createProposalForBulkUploadCsvRecord(
+        applicant.id,
+        opportunity.id,
+        `${recordNumber}`,
+      );
+      const proposalVersion = await createProposalVersionForBulkUploadCsvRecord(
+        proposal.id,
+        applicationForm.id,
+      );
+      await Promise.all(record.map<Promise<ProposalFieldValue>>(
```

Comment:
Wanted to point out that this does not guarantee sequential inserts, but I don't think that should matter since we have a `position` field.  It's more efficient this way of course. (It does complicate the tests a little bit)